### PR TITLE
backupccl, importccl: version gate backup import public schema

### PR DESCRIPTION
**Pull request by RichardJCai**
Make sure databases never have public schemas until version `PublicSchemaWithDescriptors` and ensure that this can't be circumvented by backup.

### CODE REVIEW

File: pkg/ccl/importccl/import_job.go [hunk 1]
```diff
@@ -1243,9 +1248,17 @@ func emitImportJobEvent(
...
 
 func constructSchemaAndTableKey(
-	tableDesc *descpb.TableDescriptor, schemaIDToName map[descpb.ID]string,
+	ctx context.Context,
+	tableDesc *descpb.TableDescriptor,
+	schemaIDToName map[descpb.ID]string,
+	version clusterversion.Handle,
 ) (schemaAndTableName, error) {
 	schemaName, ok := schemaIDToName[tableDesc.GetUnexposedParentSchemaID()]
+	if !version.IsActive(ctx, clusterversion.PublicSchemasWithDescriptors) {
```
Comment (adityamaru) [inline pkg/ccl/importccl/import_job.go:1257]: can we move this check above the line where we check existence in the map

**Review by adityamaru — APPROVED** [review]
Can we cleanup the commit and PR message a little bit? The import commit mentions `restore`, and the restore commit mentions `backup` which doesn't seem to have any code change?

The PR message can just be an aggregation of the two commit messages then.

File: pkg/ccl/backupccl/restore_old_versions_test.go [hunk 2]
```diff
@@ -807,3 +821,58 @@ func restorePublicSchemaRemap(exportDir string) func(t *testing.T) {
...
+		// descriptor backed public schema of database test.
+		sqlDB.Exec(t, `CREATE DATABASE test`)
+		sqlDB.Exec(t, `RESTORE d.public.t FROM $1 WITH into_db = 'test'`, LocalFoo)
+
+		row = sqlDB.QueryRow(t, `SELECT id FROM system.namespace WHERE name='test' AND "parentID"=0`)
+		var parentDBID int
+		row.Scan(&parentDBID)
+
+		row = sqlDB.QueryRow(t, fmt.Sprintf(`SELECT id FROM system.namespace WHERE name='public' AND "parentID"=%d`, parentDBID))
+		row.Scan(&publicSchemaID)
```
[the remark this answers is not in the document]
Comment (RichardJCai) [reply pkg/ccl/backupccl/restore_old_versions_test.go:874]: Added the check. The code is similar but I don't think there's much value of pulling any of it into a helper, doubt we would use the helper other than these two functions and even then the code differs in small places such that pulling it into a helper is a bit tedious.

File: pkg/ccl/importccl/import_job.go (hunk 1 above)
Comment (RichardJCai) [reply pkg/ccl/importccl/import_job.go:1257]: Done.

> Can we cleanup the commit and PR message a little bit? The import commit mentions `restore`, and the restore commit mentions `backup` which doesn't seem to have any code change?
> 
> The PR message can just be an aggregation of the two commit messages then.

Squashed the two commit messages.

**Comment by RichardJCai** [comment]
TFTR 
bors r=adityamaru

**Pull request merged**